Sources/_StringProcessing/Engine/MEBuiltins.swift

```diff
@@ -160,7 +181,46 @@ extension String {
       ? nil
       : (substr.first!, substr.endIndex)
   }
-  
+
+  /// Returns the character at `pos`, bounded by `start`, as well as the upper
+  /// boundary of the returned character.
+  ///
+  /// This function handles loading a character from a string while respecting
+  /// an start boundary, even if that start boundary is sub-character or sub-scalar.
+  ///
+  ///   - If `pos` is at or past `start`, this function returns `nil`.
+  ///   - If `start` is between `pos` and the next grapheme cluster boundary (i.e.,
+  ///     `start` is before `self.index(after: pos)`, then the returned character
+  ///     is smaller than the one that would be produced by `self[pos]` and the
+  ///     returned index is at the start of that character.
+  ///   - If `start` is between `pos` and the next grapheme cluster boundary, and
+  ///     is not on a Unicode scalar boundary, the partial scalar is dropped. This
+  ///     can result in a `nil` return or a character that includes only part of
+  ///     the `self[pos]` character.
+  ///
+  /// - Parameters:
+  ///   - pos: The position to load a character from.
+  ///   - start: The limit for the character at `pos`.
+  /// - Returns: The character at `pos`, bounded by `start`, if it exists, along
+  ///   with the upper bound of that character. The upper bound is always
+  ///   scalar-aligned.
+  func characterAndStart(at pos: String.Index, limitedBy start: String.Index) -> (Character, String.Index)? {
```

Comment:
It's actually the lower bound of the character. Let's add a label to the return tuple value that's something like:
`(Character, characterStart: String.Index)`